fem/libceed/operator.hpp

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) 2010-2020, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_LIBCEED_OPERATOR
+#define MFEM_LIBCEED_OPERATOR
+
+#include "../../config/config.hpp"
+
+#ifdef MFEM_USE_CEED
+#include "ceed.hpp"
+#include <ceed.h>
+#include "../../linalg/operator.hpp"
+#include "solvers-utility.h"
+
+namespace mfem
+{
+
+/** A base class to represent a CeedOperator as an MFEM Operator. */
+class MFEMCeedOperator : public Operator
+{
+protected:
+   CeedOperator oper;
+   bool owns_oper;
+   CeedVector u, v;
+
+   MFEMCeedOperator() : oper(nullptr), owns_oper(true), u(nullptr),
+      v(nullptr) { }
+
+public:
+   void Mult(const Vector &x, Vector &y) const;
+   void AddMult(const Vector &x, Vector &y) const;
+   void GetDiagonal(Vector &diag) const;
+   virtual ~MFEMCeedOperator()
+   {
+      if (owns_oper) { CeedOperatorDestroy(&oper); }
+      CeedVectorDestroy(&u);
+      CeedVectorDestroy(&v);
+   }
+   CeedOperator GetCeedOperator() const { return oper; }
+};
+
+class UnconstrainedMFEMCeedOperator : public MFEMCeedOperator
+{
+public:
+   UnconstrainedMFEMCeedOperator(CeedOperator ceed_op)
+   {
+      oper = ceed_op;
+      owns_oper = false;
+      CeedElemRestriction er;
+      CeedOperatorGetActiveElemRestriction(oper, &er);
+      int s;
+      CeedElemRestrictionGetLVectorSize(er, &s);
+      height = width = s;
+      CeedVectorCreate(internal::ceed, height, &v);
+      CeedVectorCreate(internal::ceed, width, &u);
+   }
+
+   Operator * SetupRAP(const Operator *Pi, const Operator *Po)
+   {
+      return Operator::SetupRAP(Pi, Po);
+   }
+};
+
+/** Wraps a CeedOperator in an mfem::Operator, with essential boundary
+    conditions. */
+class ConstrainedMFEMCeedOperator : public Operator
```

Comment:
Shouldn't this class inherit from `ConstrainedOperator`?

Reply:
Maybe, but the `ConstrainedOperator` constructor needs an `mfem::Operator` as an argument, so if I want to construct that MFEM operator here from a given `CeedOperator` I can't use inheritance. I am open to other designs but I don't think what you propose is a simple change.

Reply:
An `MFEMCeedOperator` is an `mfem::Operator`, wouldn't that work?

Reply:
I don't understand. Are you proposing rewriting the `ConstrainedMFEMCeedOperator` constructor to take an `mfem::Operator` instead of a `CeedOperator` argument? I want the user to be able to just call this once and not construct a whole chain of objects.

Reply:
I'm just worried that we might want to use a `ConstrainedMFEMCeedOperator` *as a* `ConstrainedOperator`. If you tell me that this will never be the case because this will never be used outside the libCEED integration then I'm fine with it.